### ISSUE TITLE
Packets are too closely tied to their format - store packet info in the libtrace packet struct?

### ISSUE THREAD

**Issue by rsanger** [wanduow/libtrace#28]
TLDR: Store packet information (timestamp, direction, size etc.) in the libtrace packet header rather than in the format's packet header, otherwise we are relying on format support.

Libtrace formats all support different fields with some being more restrictive than others, a libtrace packet is tied to a specific format. Calling the set and get functions on a packet call the format specific functions which store this information in the packet format header (included in the packet buffer).

Packets read from a format without a field cannot be set and then stored to a format with support. The results in odd behaviour, for example running:
tracemerge --set-interface erf:output.erf erf:in1.erf erf:in2.erf
Will result in a unique direction/port assigned to in1.erf and in2.erf in the merged result output.erf.
However running the same command with pcap files as input and erf as output will not, despite the erf format supporting the direction field.

Most of these fields are already cached in the libtrace packet, so this would not drastically increase its size. Logic would need to be updated to allow writing to these without format support, and deal with cases such as the format only providing a get and not a set.

It might be best to pre-emptively fill all these cached values when a packet is read, I suspect this would end up faster. When getting a field this would remove the need to check for and then call the format specific get option. Another consideration would be ensuring that setting a field gets written to the format header, so that writing back to the same format type wont lose the updated field. I'm certain there will be more edge cases here.

Other advantages include:
As a result user constructed packets no longer would need to be tied to a 'dead format'.
The resulting packet could also then be used as the rt header.

**Comment by rsanger** [comment]
We should remove the need for promote_packet() see cd8cd74 with this resolution to this issue.

One possibility would be to do this only when setting a field but not when reading from the format:
 * Make all trace_set_*(packet) calls write to the packet cache, not the packet header
 * Require formats read from the cache even when writing back their own packet to get the new values
 * Consider all cases where the packet cache is cleared and this information might be lost
